From 51b63835a99d660933546251e45ce499500ff355 Mon Sep 17 00:00:00 2001
From: Janne Grunau <j@jannau.net>
Date: Sun, 24 May 2026 00:04:57 +0200
Subject: [PATCH 1/1] power: domain: apple: Ignore power states with auto
 enable flag

The current code timeouts while trying to enable the "apcie_sys_st" NVMe
power states on Apple M3 Pro/Max SoCs. Ignoring these power states which
are marked as auto enable avoids the timeout with a working NVMe.
The Linux driver clears the auto enable bit but preserving the initial
boot state from Apple's iboot/mboot in u-boot seems preferable.

Signed-off-by: Janne Grunau <j@jannau.net>
---
 drivers/power/domain/apple-pmgr.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/power/domain/apple-pmgr.c b/drivers/power/domain/apple-pmgr.c
index 71fedb405da..3f74b5a277a 100644
--- a/drivers/power/domain/apple-pmgr.c
+++ b/drivers/power/domain/apple-pmgr.c
@@ -14,6 +14,8 @@
 #include <syscon.h>
 
 #define APPLE_PMGR_RESET	BIT(31)
+#define APPLE_PMGR_AUTO_ENABLE	BIT(28)
+#define APPLE_PMGR_PS_RESET	BIT(12)
 #define APPLE_PMGR_DEV_DISABLE	BIT(10)
 #define APPLE_PMGR_WAS_CLKGATED	BIT(9)
 #define APPLE_PMGR_WAS_PWRGATED BIT(8)
@@ -83,6 +85,14 @@ static int apple_pmgr_ps_set(struct power_domain *power_domain, u32 pstate)
 {
 	struct apple_pmgr_priv *priv = dev_get_priv(power_domain->dev);
 	uint reg;
+	int ret;
+
+	ret = regmap_read(priv->regmap, priv->offset, &reg);
+	if (ret < 0)
+		return ret;
+
+	if (FIELD_GET(APPLE_PMGR_AUTO_ENABLE, reg))
+		return 0;
 
 	regmap_update_bits(priv->regmap, priv->offset, APPLE_PMGR_PS_TARGET,
 			   FIELD_PREP(APPLE_PMGR_PS_TARGET, pstate));
